files/utils.cc

```diff
@@ -844,18 +841,17 @@ void U7copy(
 		throw;
 	}
 	size_t bufsize = 0x8000;
-	char *buf = new char[0x8000];
+	std::vector<char> buf(0x8000);
 	in.seekg(0, ios::end);      // Get filesize.
 	size_t filesize = in.tellg();
 	in.seekg(0, ios::beg);
 	while (filesize > 0) {      // Copy.
 		size_t toread = bufsize < filesize ? bufsize : filesize;
-		in.read(buf, toread);
-		out.write(buf, toread);
+		in.read(&buf[0], toread);
+		out.write(&buf[0], toread);
 		filesize -= toread;
 	}
```

Comment:
I was thinking that swapping this by `out << in.rdbuf();` is probably going to be much faster. And it would avoid the undefined behavior from the way the code obtains the file size.

Reply:
OK, I didn't know the `rdbuf` trick, I was actually thinking about using ostreamuf/istreamubf iterators.
Where's the UB, BTW?

Reply:
There are two different sources of UB, both of which come from the C underbelly:

- for a text file, ftell returns a token value that can be used to seek to the same position, but which need not be related to the byte position in the file;
- for binary mode files, ftell does return a byte offset, but fseek to the end is not guaranteed to go one past the last byte of the file.

A couple references:

- [FIO19-C](https://wiki.sei.cmu.edu/confluence/display/c/FIO19-C.+Do+not+use+fseek%28%29+and+ftell%28%29+to+compute+the+size+of+a+regular+file)
- [blog post on Explicit C++](https://web.archive.org/web/20180314195042/http://cpp.indi.frih.net/blog/2014/09/how-to-read-an-entire-file-into-memory-in-cpp/)

And using the `<<` with `rdbuf()` is a LOT faster because it does not behave like formatted output at all; rather, it uses a specialized overload of `<<` which directly uses the underlying stream buffer objects to copy data as fast as possible.

Reply:
"Love" those caveats. Even more with C, as in C++ it's at least common to expect UB when doing anything ;)